### ISSUE TITLE
Ruby: oneof fields containing a default-value enum don't preserve presence

### ISSUE THREAD

**Issue by kmontag** [google/protobuf#1204]
I'm trying to use a `oneof` wrapper to add presence information to an `enum` field, but when the field is set to its default value, it encodes as if it were never set at all. Perhaps better explained by example:

```ruby
require 'google/protobuf'

Google::Protobuf::DescriptorPool.generated_pool.build do
  add_message "TestMessage" do
    oneof :wrapper do
      optional :value, :enum, 1, "TestEnum"
    end
  end
  add_enum "TestEnum" do
    value :FOO, 0
    value :BAR, 1
  end
end

TestMessage = Google::Protobuf::DescriptorPool.generated_pool.lookup("TestMessage").msgclass
TestEnum = Google::Protobuf::DescriptorPool.generated_pool.lookup("TestEnum").enummodule

message = TestMessage.new

# Non-default value
message.value = :BAR
puts message.wrapper == :value # => true
puts message.value == :BAR # => true
puts TestMessage.decode(TestMessage.encode message).wrapper == :value # => true
puts TestMessage.decode(TestMessage.encode message).value == :BAR # => true

# Default value
message.value = :FOO
puts message.wrapper == :value # => true
puts message.value == :FOO # => true
puts TestMessage.decode(TestMessage.encode message).wrapper == :value # => false (unexpected)
puts TestMessage.decode(TestMessage.encode message).wrapper == nil # => true (unexpected)
puts TestMessage.decode(TestMessage.encode message).value == :FOO # => false (unexpected)
puts TestMessage.decode(TestMessage.encode message).value == nil # => true (unexpected)
```

I haven't tested in other languages, so not sure if this is an issue outside of Ruby as well.

**Comment by haberman** [comment]
Confirmed -- thanks for the report! I checked and C++ does not have this issue. This should be fixed.

**Comment by haberman** [comment]
We should also add this case to the conformance tests, since it's easy to get wrong.

**Comment by blowmage** [comment]
@haberman Has this been fixed? I also ran into this. Here is [my reproduction](https://gist.github.com/blowmage/d90f1428f0f54b25f682).

**Comment by blowmage** [comment]
Here is [another reproduction](https://gist.github.com/blowmage/f9b54d7463b7d2cb4245). This is a pretty major bug for us.

/cc @jgeewax @pcostell

**Comment by blowmage** [comment]
FWIW, here is a [similar issue](https://github.com/dcodeIO/protobuf.js/issues/329) filed against protobuf.js.

**Comment by pcostell** [comment]
I tried updating some of the protobuf tests and can only get this to fail with jruby:
```

rvm ruby-2.3.0
rake test  <== doesn't fail oneof tests

rvm jruby
rake test <== encode then decode fails to pick up oneof
```

**Comment by blowmage** [comment]
The GRPC gem we are using defines a `~> 3.0.0.alpha.5.0.2` dependency on google-protobuf. I've been running against the `3.0.0.alpha.5.0.3` release.

**Comment by blowmage** [comment]
I built a gem from the master branch and installed it locally. I do not have this `oneof` issue using that gem on MRI. I have not tried jRuby.

**Comment by haberman** [comment]
Yes this was fixed in https://github.com/google/protobuf/pull/1139 -- sorry I forgot to mention this in the commit message.

I don't remember off the top of my head whether the latest RubyGems release includes this fix or not.

**Comment by blowmage** [comment]
The commits are dated a week after the last release. Is there an ETA for the next release?

**Comment by blowmage** [comment]
@haberman The google-protobuf gem version 3.0.0.alpha.5.0.4 released yesterday does *not* include this fix.

**Comment by blowmage** [comment]
@haberman Ping. This fix is on master, but not in the released rubygem. Would it be possible to get a rubygem release that includes this fix?

**Comment by haberman** [comment]
Sorry for the delay on this! There were some issues with JSON support that I wanted to resolve before we cut another release. I had updated JSON to support proper proto3 camelCased names, but I had to do some more work to make sure this wouldn't break existing users.

In any case, everything is ready for us to cut a new patch release now. @nicolasnoble will be doing the actual release.

**Comment by blowmage** [comment]
Wonderful! Thanks for the update!

**Comment by quartzmo** [comment]
@haberman @nicolasnoble Fantastic news! Thank you!

**Comment by blowmage** [comment]
@nicolasnoble Is there an ETA on the next ruby release that will contain this fix? Is there a branch that we can use to work against the code that will be in the release?

**Comment by blowmage** [comment]
Ping.

**Comment by haberman** [comment]
Updated Gems are on RubyGems. This is fixed in 3.0.0.alpha.5.0.5. Sorry for the long wait and thanks for your patience.

**Comment by quartzmo** [comment]
@haberman Great news! Thank you!

**Comment by blowmage** [comment]
Fantastic! Thanks!

**Comment by blowmage** [comment]
The fix for this was shipped with google-protobuf version 3.0.0.alpha.5.0.5. Thanks all!
